### PR TITLE
Convert remaining JS/JSX files to TypeScript

### LINKED ISSUE #1142: Convert reducer files from JS to TS
Reference PR
https://github.com/mattermost/mattermost-plugin-jira/pull/1140

### PR DESCRIPTION
**Pull request by rafaumeu**
## Summary

Converts all remaining `.js` and `.jsx` files to TypeScript (`.ts`/`..`tsx`), following the pattern from PR #1140.

## Files Converted (11 total)

**Utils:**
- `utils/posts.jsx` -> `posts.ts`
- `utils/user_agent.jsx` -> `user_agent.ts`

**Components:**
- `components/setting.jsx` -> `setting.tsx`
- `components/form_button.jsx` -> `form_button.tsx`
- `components/input.jsx` -> `input.tsx`
- `components/jira_field.jsx` -> `jira_field.tsx`
- `components/setup_ui/setup_ui.jsx` -> `setup_ui.tsx`
- `components/setup_ui/index.js` -> `index.ts`
- `components/jira_issue_selector/jira_issue_selector.jsx` -> `jira_issue_selector.tsx`
- `components/jira_issue_selector/index.js` -> `index.ts`
- `components/modals/full_screen_modal/full_screen_modal.jsx` -> `full_screen_modal.tsx`

## Changes

- Replaced `PropTypes` with proper TypeScript `type`/`interface` definitions
- Added typed function parameters and return types
- Updated event handler types (e.g., `ChangeEvent`, `KeyboardEvent`)
- Updated import paths to remove `.jsx` extensions
- `disabled` prop on AsyncSelect changed to `isDisabled` (react-select v3 API)

## Ticket Link

Fixes #1142

## Checklist

- [x] Converted all .js and .jsx files in webapp/src to .ts/.tsx
- [x] Replaced PropTypes with TypeScript types
- [x] Maintained original component behavior
- [x] `te char/e2e` tests not affected (no logic changes)

`/release-note`

## Change Impact: 🟡 Medium

**Reasoning:** This PR converts 11 files from JavaScript/PropTypes to TypeScript across components and utilities. While most changes are typing-only migrations following an established pattern (PR `#1140`), there are some behavioral modifications (AsyncSelect `disabled` → `isDisabled` API update, Input component validation state clearing, numeric input handling) that introduce regression risk across multiple component implementations.

**Regression Risk:** Moderate. The PR maintains intended behavior per the description, but affects multiple component files (FormButton, Input, JiraField, JiraIssueSelector, FullScreenModal, Setting, SetupUI) with behavioral refinements beyond pure typing. The AsyncSelect API change and Input component's modified `componentDidUpdate` hook clearing invalid state when props change could have unexpected interactions in parent components if not thoroughly tested. Utilities (posts.ts, user_agent.ts) are low-risk but shared. Existing e2e/char tests are expected to remain unaffected, but edge cases in numeric input handling and validation state management warrant attention.

**QA Recommendation:** Manual QA is recommended with medium-to-high priority. Test focus should include: (1) AsyncSelect components with dynamic enable/disable scenarios in Jira-related flows; (2) numeric input edge cases and validation state transitions; (3) FullScreenModal keyboard event handling; (4) Jira field selection workflows. Automated test coverage appears adequate (PR notes existing tests remain unaffected), but behavioral modifications and integration points between modified components and parent consumers should be validated in staging/development environments.

*Generated by CodeRabbitAI*

### CODE REVIEW

File: webapp/src/components/jira_field.tsx [hunk 1]
```diff
@@ -2,7 +2,7 @@
 // See LICENSE.txt for license information.
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import type {Theme} from 'mattermost-redux/types/preferences';
```
Comment (nang2049) [inline webapp/src/components/jira_field.tsx]: The Theme type lives in mattermost-redux/selectors/entities/preferences, which is what every other file in this webapp uses. The build passes only because import type is erased before webpack runs.

```
import type {Theme} from 'mattermost-redux/selectors/entities/preferences';
```

File: webapp/src/components/jira_issue_selector/jira_issue_selector.tsx [hunk 2]
```diff
@@ -2,7 +2,7 @@
 // See LICENSE.txt for license information.
 
 import React, {Component} from 'react';
-import PropTypes from 'prop-types';
+import type {Theme} from 'mattermost-redux/types/preferences';
```
Comment (nang2049) [inline webapp/src/components/jira_issue_selector/jira_issue_selector.tsx]: Same as in jira_field.tsx

File: webapp/src/components/jira_field.tsx [hunk 3]
```diff
@@ -18,29 +18,68 @@ import JiraUserSelector from './data_selectors/jira_user_selector';
...
+    schema: FieldSchema;
+    allowedValues?: AllowedValue[];
+};
+
+type Props = {
+    id: string;
+    instanceID: string;
+    field: Field;
+    projectKey: string;
+    issueMetadata: Record<string, any>;
```
[the remark this answers is not in the document]
Comment (nang2049) [reply webapp/src/components/jira_field.tsx]: Agreed with this, the caller in jira_fields.tsx passes `projectKey: string | undefined` and `issueMetadata: IssueMetadata | null`. This component types them as string and `Record<string, any>` the second of which then propagates straight into `<JiraEpicSelector issueMetadata={...} />` which expects IssueMetadata.

File: webapp/src/components/jira_field.tsx [hunk 4]
```diff
@@ -18,29 +18,68 @@ import JiraUserSelector from './data_selectors/jira_user_selector';
...
-        projectKey: PropTypes.string.isRequired,
-        issueMetadata: PropTypes.object.isRequired,
-        obeyRequired: PropTypes.bool,
-        onChange: PropTypes.func.isRequired,
-        value: PropTypes.any,
-        isFilter: PropTypes.bool,
-        theme: PropTypes.object.isRequired,
-        addValidate: PropTypes.func.isRequired,
-        removeValidate: PropTypes.func.isRequired,
+type JiraFieldData = {
```
Comment (nang2049) [inline webapp/src/components/jira_field.tsx:21]: JiraField, IssueMetadata, etc. are already exported from types/model. The local Field, FieldSchema, AllowedValue types are structurally narrower than the model's JiraField, which is why isTeamField(field) would fail typechecking, isTeamField is typed against JiraField | FilterField not this local shape. Use the canonical model types so the rest of the codebase stays consistent.

File: webapp/src/components/jira_field.tsx [hunk 5]
```diff
@@ -111,7 +150,7 @@ export default class JiraField extends React.Component {
                 <JiraEpicSelector
```
Comment (nang2049) [inline webapp/src/components/jira_field.tsx:122]: on line 93 style.getStyle() only returns { jiraIcon: {...} } so selectComponent is undefined and this is a style={undefined} no-op. The original .jsx had the same bug but it was hidden because JSX didn't typecheck. Either restore the missing entry or drop the style prop.

**Review by nang2049 — COMMENTED** [review]
Thanks @rafaumeu I have a few comments :)

**Comment by rafaumeu** [comment]
Review feedback addressed in commit ce2932c:

**input.tsx**
- `parseInt()` replaced with `Number()` — preserves decimal values instead of truncating to integers
- `isValid()` no longer rejects `0` as invalid — now explicitly checks for `undefined`/`null`/`''` instead of relying on truthiness
- Value rendering uses `??` instead of `||` so `0` displays correctly in the input

**jira_field.tsx**
- `Theme` import corrected to `mattermost-redux/selectors/entities/preferences` (canonical export path)
- Removed local `Field`, `FieldSchema`, and `AllowedValue` type definitions — now imports canonical `JiraField`, `AllowedValue` from `types/model`. This eliminates type narrowing mismatches with the rest of the codebase.
- `Props` type updated to match the actual caller (`jira_fields.tsx`): `projectKey?: string` (optional) and `issueMetadata: IssueMetadata | null` instead of `Record<string, any>`
- Removed dead `selectComponent` style prop from `IconOption` — `getStyle()` only returns `jiraIcon`, so the `style` prop on `components.Option` was a no-op

**jira_issue_selector.tsx**
- Same `Theme` import path fix

**Comment by rafaumeu** [comment]
Hi @nang2049, I've addressed all review feedback in commit ce2932c:

1. **Theme import** — Fixed in both `jira_field.tsx` and `jira_issue_selector.tsx`: now imports from `mattermost-redux/selectors/entities/preferences` (canonical path used across the webapp).
2. **Canonical types** — Removed local `Field`, `FieldSchema`, and `AllowedValue` type definitions. Now imports `JiraField`, `IssueMetadata`, and `AllowedValue` directly from `types/model`, which fixes the type narrowing issue with `isTeamField()`.
3. **Props alignment** — `projectKey` is now `string | undefined` and `issueMetadata` is `IssueMetadata | null`, matching what `jira_fields.tsx` actually passes.
4. **Removed dead style** — Dropped `style={getStyle().selectComponent}` (was always `undefined`).
5. **`input.tsx`** — Replaced `parseInt` with `Number()` to preserve decimal values for float fields.

@coderabbitai could you re-review?
